src/background/services/blacklist.js

```diff
@@ -0,0 +1,117 @@
+//heavily ai generated this file can be optimized alot
+import {parse} from 'tldts';
+
+const OISD_BLOCKLIST_URL = 'https://big.oisd.nl/domainswild';
+
+let oisdBlocklist = new Set();
+let oisdRegexList = [];
+let isBlocklistLoaded = false;
+let blocklistLoadPromise = null;
+
+/**
+ * Load the OISD blocklist (domains or regexes) exactly once.
+ * @returns {Promise<{ domains: Set<string>, regexes: RegExp[] }>}
+ */
+export function loadBlocklist() {
+  if (blocklistLoadPromise) {
+    return blocklistLoadPromise;
+  }
+  if (isBlocklistLoaded) {
+    return Promise.resolve({domains: oisdBlocklist, regexes: oisdRegexList});
+  }
+
+  blocklistLoadPromise = (async () => {
+    try {
+      const response = await fetch(OISD_BLOCKLIST_URL);
+      if (!response.ok) {
+        throw new Error(`HTTP ${response.status}`);
+      }
+      const text = await response.text();
+      const lines = text.split('\n').map((l) => l.trim());
+
+      // Detect if file is in regex format
+      const isRegexFormat = lines.some((l) => l.startsWith('# Syntax: Regex'));
+      if (isRegexFormat) {
+        oisdRegexList = lines
+          .filter((l) => l && !l.startsWith('#'))
+          .map((l) => {
+            try {
+              // strip leading/trailing slash and flags
+              const body = l.replace(/^\/|\/[gimusy]*$/g, '');
+              return new RegExp(body);
+            } catch {
+              return null;
+            }
+          })
+          .filter((rx) => rx);
+      } else {
+        oisdBlocklist = new Set(
+          lines
+            .filter((l) => l && !/^[#!/\s]/.test(l))
+            .map((l) => l.toLowerCase().replace(/^[*.]+/, ''))
+            .filter(
+              (domain) =>
+                /^[a-z0-9.-]+$/.test(domain) && !/^\d+(\.\d+){3}$/.test(domain),
+            ),
+        );
+      }
+
+      isBlocklistLoaded = true;
+      return {domains: oisdBlocklist, regexes: oisdRegexList};
+    } catch (err) {
+      console.error('[blacklist] load failed', err);
+      // Reset so we can retry later if needed
+      isBlocklistLoaded = false;
+      oisdBlocklist.clear();
+      oisdRegexList = [];
+      blocklistLoadPromise = null;
+      return {domains: oisdBlocklist, regexes: oisdRegexList};
+    }
+  })();
+
+  return blocklistLoadPromise;
+}
+
+// Kick off initial load
+loadBlocklist();
```

Comment:
why calling it here? the function is exported and the part importing it should have control on the initial load.